src/NUnitFramework/framework/Constraints/DictionaryContainsKeyConstraint.cs

```diff
@@ -66,31 +72,242 @@ public override string Description
+        [Obsolete("DictionaryContainsKeyConstraint now uses the comparer which the dictionary is based on. To test using a comparer which the dictionary is not based on, use a collection constraint on the set of keys.")]
+        public new CollectionItemsEqualConstraint Using<T>(IEqualityComparer<T> comparer)
+        {
+            _isDeprecatedMode = true;
+            return base.Using(comparer);
+        }
+
+        /// <summary>
+        /// Flag the constraint to use the supplied boolean-returning delegate.
+        /// </summary>
+        /// <param name="comparer">The supplied boolean-returning delegate to use.</param>
+        [Obsolete("DictionaryContainsKeyConstraint now uses the comparer which the dictionary is based on. To test using a comparer which the dictionary is not based on, use a collection constraint on the set of keys.")]
+        public new CollectionItemsEqualConstraint Using<T>(Func<T, T, bool> comparer)
+        {
+            _isDeprecatedMode = true;
+            return base.Using(comparer);
+        }
+
+        #endregion
+
+        private MethodInfo GetContainsKeyMethod(object keyedItemContainer)
+        {
+            if (keyedItemContainer == null) throw new ArgumentNullException(nameof(keyedItemContainer));
+            var instanceType = keyedItemContainer.GetType();
+
+            var method = FindContainsKeyMethod(instanceType)
+                         ?? instanceType
+                            .GetInterfaces()
+                            .Concat(GetBaseTypes(instanceType))
+                            .Select(FindContainsKeyMethod)
+                            .FirstOrDefault(m => m != null);
+
+            return method;
+        }
+
+        private MethodInfo FindContainsKeyMethod(Type type)
+        {
+            var methods = type.GetMethods(BindingFlags.Instance | BindingFlags.Public);
+            var method = methods.FirstOrDefault(m =>
+                m.ReturnType == typeof(bool)
+                && m.Name == "ContainsKey"
+                && m.GetParameters().Length == 1);
+
+            if (method == null && type.GetTypeInfo().IsGenericType)
+            {
+                var definition = type.GetGenericTypeDefinition();
+                var tKeyGenericArg = definition.GetGenericArguments().FirstOrDefault(typeArg => typeArg.Name == "TKey");
+
+                if (tKeyGenericArg != null)
+                {
+                    method = definition
+                             .GetMethods(BindingFlags.Instance | BindingFlags.Public)
+                             .FirstOrDefault(m =>
+                                 m.ReturnType == typeof(bool)
+                                 && m.Name == "Contains"
+                                 && m.GetParameters().FirstOrDefault()?.ParameterType == tKeyGenericArg);
+
+                    if (method != null)
+                    {
+#if NETSTANDARD1_4
+                        method = methods.Single(m => Matched(method, m));                    
```

Comment:
As far as I can see, .NET Standard 1.4 just puts us in a really uncomfortable spot.

I'd take a different approach. I wouldn't even bother doing `var definition = type.GetGenericTypeDefinition();` on .NET Standard 1.4. May as well copy the whole block and have .NET Standard 1.4 operate on the instantiated `type` instead of `definition`. Would that make sense?

Reply:
I can help with the coding if that's clearer than me trying to explain it. :D